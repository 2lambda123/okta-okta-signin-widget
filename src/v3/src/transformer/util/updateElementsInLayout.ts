/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  StepperLayout,
  UISchemaElement,
  UISchemaLayout,
  UISchemaLayoutType,
} from '../../types';

type PredicateFn = (uischema: UISchemaElement) => boolean;
type MapFn = (uischema: UISchemaElement) => UISchemaElement;
type CallbackFn = (uischema: UISchemaElement) => void;

type Options = {
  layout: UISchemaLayout;
  predicate: PredicateFn;
  callback: CallbackFn;
};

type OptionsOld = {
  layout: UISchemaLayout;
  mapFn: MapFn;
  predicateFn: PredicateFn;
};

export const updateElementsInLayout = (options: OptionsOld) => {
  const fn = (
    layout: UISchemaLayout,
    mapFn: MapFn,
    predicateFn: PredicateFn = () => true,
  ): UISchemaLayout => {
    // eslint-disable-next-line no-param-reassign
    layout.elements = layout.elements.map((element) => {
      const { type } = element;
      if (type === UISchemaLayoutType.STEPPER) {
        // eslint-disable-next-line no-param-reassign
        (element as StepperLayout).elements = (element as StepperLayout).elements
          .map((el) => fn(el, mapFn, predicateFn));
        return element;
      }

      if ([UISchemaLayoutType.HORIZONTAL, UISchemaLayoutType.VERTICAL]
        .includes(type as UISchemaLayoutType)) {
        return fn(element as UISchemaLayout, mapFn, predicateFn);
      }

      if (predicateFn(element)) {
        return mapFn(element);
      }

      return element;
    });

    return layout;
  };

  return fn(options.layout, options.mapFn, options.predicateFn);
};

export const traveseLayout = (options: Options) => {
  const fn = (
    layout: UISchemaLayout,
    predicateFn: PredicateFn,
    callback: CallbackFn,
  ) => {
    layout.elements.forEach((element) => {
      if (predicateFn(element)) {
        callback(element);
      }

      const { type } = element;

      if (type === UISchemaLayoutType.STEPPER) {
        (element as StepperLayout).elements
          .forEach((el) => fn(el, predicateFn, callback));
        return;
      }

      if ([UISchemaLayoutType.HORIZONTAL, UISchemaLayoutType.VERTICAL]
        .includes(type as UISchemaLayoutType)) {
        fn(element as UISchemaLayout, predicateFn, callback);
        return;
      }
    });
  };

  fn(options.layout, options.predicate, options.callback);
}
