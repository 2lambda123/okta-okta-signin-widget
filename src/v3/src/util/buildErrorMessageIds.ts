/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export const buildErrorMessageIds = (errors: string[], fieldName: string): string => {
  if (errors.length === 1) {
    return `${fieldName}-error`;
  }
  const ids = errors.map((_: string, index: number) => `${fieldName}-error-${index}`);
  return ids.join(' ');
};
