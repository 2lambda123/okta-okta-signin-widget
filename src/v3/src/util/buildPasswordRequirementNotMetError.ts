/*
 * Copyright (c) 2023-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { ListItem, PasswordValidation } from '../types';
import { loc } from './locUtil';

export const buildPasswordRequirementNotMetError = (
  passwordRequirements: ListItem[],
  passwordValidations: PasswordValidation,
): string | undefined => {
  // TODO: Add ticket here to create new translation key
  const errorMessagePrefix = loc('registration.error.password.passwordRequirementsNotMet.prefix', 'login');
  const errorMessageSuffix = passwordRequirements.map((requirement: ListItem) => {
    const ruleValue = passwordValidations[requirement.ruleKey];
    return ruleValue === false ? requirement.label : undefined;
  }).filter(Boolean).join(', ') || undefined;

  return typeof errorMessageSuffix !== 'undefined'
    ? `${errorMessagePrefix} ${errorMessageSuffix}`
    : undefined;
};
