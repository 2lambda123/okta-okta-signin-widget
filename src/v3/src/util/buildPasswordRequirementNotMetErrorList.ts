/*
 * Copyright (c) 2023-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxMessageWithName, ListItem, PasswordValidation } from '../types';
import { loc } from './locUtil';

export const buildPasswordRequirementNotMetErrorList = (
  passwordRequirements: ListItem[],
  passwordValidations: PasswordValidation,
  fieldName: string,
): IdxMessageWithName[] => {
  const errorMessages: IdxMessageWithName[] = [];

  passwordRequirements.forEach((requirement: ListItem) => {
    const ruleValue = passwordValidations[requirement.ruleKey];
    if (ruleValue === false) {
      errorMessages.push({
        name: fieldName,
        class: 'ERROR',
        message: requirement.label,
        i18n: { key: '' },
      });
    }
  });

  if (errorMessages.length > 0) {
    // TODO: Add ticket here to create new translation key that includes colon
    errorMessages.unshift({
      name: fieldName,
      class: 'ERROR',
      message: loc('registration.error.password.passwordRequirementsNotMet', 'login'),
      i18n: { key: 'registration.error.password.passwordRequirementsNotMet' },
    });
  }

  return errorMessages;
};
