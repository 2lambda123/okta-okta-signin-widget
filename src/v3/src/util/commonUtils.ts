/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

// eslint-disable-next-line
export const getDisplayName = (WrappedComponent: any): string => (
  WrappedComponent.displayName || WrappedComponent.name || 'Component'
);

export const handleFormFieldChange = (
  setDirtyFn: ((isDirty: boolean) => void) | undefined,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  handleChange: (path: string, value: any) => void,
  path: string,
  // eslint-disable-next-line
  value: any,
): void => {
  setDirtyFn?.(true);
  handleChange(path, value);
};
