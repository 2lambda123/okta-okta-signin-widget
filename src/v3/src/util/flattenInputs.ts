/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { Input } from '@okta/okta-auth-js';

export const flattenInputs = (input: Input, name = ''): Input[] => {
  const res: Input[] = [];
  const { value } = input;

  if (Array.isArray(value)) {
    return value.reduce((acc, curr) => [
      ...acc,
      ...flattenInputs(curr, input.name),
    ], res);
  }

  return [{
    ...input,
    name: name ? `${name}.${input.name}` : input.name,
  }];
};
