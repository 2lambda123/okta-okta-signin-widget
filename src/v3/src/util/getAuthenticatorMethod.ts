/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

export const getAuthenticatorMethod = (
  transaction: IdxTransaction,
): string | undefined => {
  // currentAuthenticator is from enrollment flows and currentAuthenticatorEnrollment is from verify flows
  // @ts-ignore Add currentAuthenticatorEnrollment to IdxContext type
  const { context: { currentAuthenticator, currentAuthenticatorEnrollment } } = transaction;
  const enrollMethods = currentAuthenticator?.value?.methods;
  const verifyMethods = currentAuthenticatorEnrollment?.value?.methods;

  return (enrollMethods && enrollMethods[0].type) || (verifyMethods && verifyMethods[0].type);
};
