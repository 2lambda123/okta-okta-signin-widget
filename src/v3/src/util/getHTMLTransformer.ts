/*
 * Copyright (c) 2023-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { convertNodeToElement, Options } from 'react-html-parser';

export const getHTMLTransformer = (
  currentNodeName: string,
  newNodeName: string,
  attributes?: Record<string, unknown>,
): Options['transform'] => {
  const transform: Options['transform'] = (node, index) => {
    if (node.type === 'tag' && node.name === currentNodeName) {
      // eslint-disable-next-line no-param-reassign
      node.name = newNodeName;
      if (typeof attributes !== 'undefined') {
        // eslint-disable-next-line no-param-reassign
        node.attribs = attributes;
      }
      return convertNodeToElement(node, index, transform!);
    }
    return undefined;
  };

  return transform;
};
