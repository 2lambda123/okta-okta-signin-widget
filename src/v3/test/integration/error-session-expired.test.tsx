/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { waitFor } from '@testing-library/preact';
import { setup } from './util';

import mockResponse from '../../src/mocks/response/idp/idx/identify/error-session-expired.json';
import identifyWithPassword from '../../src/mocks/response/idp/idx/introspect/default.json';

describe('error-session-expired', () => {
  it('should render form', async () => {
    const { container, findByText } = await setup({ mockResponse });
    await findByText(/You have been logged out due to inactivity. Refresh or return to the sign in screen./);
    expect(container).toMatchSnapshot();
  });

  it('should bootstrap widget when clicking "Back to sign in" link', async () => {
    const {
      user,
      findByText,
      authClient,
    } = await setup({
      mockResponses: {
        '/introspect': {
          data: mockResponse,
          status: 200,
        },
        '/idp/idx/introspect': {
          data: identifyWithPassword,
          status: 200,
        },
      },
    });

    const cancelLink = await findByText(/Back to sign in/);
    await user.click(cancelLink);

    expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
      'POST',
      'https://oie-123456.oktapreview.com/idp/idx/introspect',
      {
        data: {
          interactionHandle: 'fake-interactionhandle',
        },
        headers: {
          Accept: 'application/ion+json; okta-version=1.0.0',
          'Content-Type': 'application/ion+json; okta-version=1.0.0',
          'X-Okta-User-Agent-Extended': 'okta-auth-js/9.9.9',
        },
        withCredentials: true,
      },
    );
  });

  it('should have focus on "Back to sign in" link', async () => {
    const { findByText } = await setup({ mockResponse });
    await findByText(/You have been logged out due to inactivity. Refresh or return to the sign in screen./);
    const cancelLink = await findByText(/Back to sign in/);
    await waitFor(() => expect(cancelLink).toHaveFocus());
  });
});
