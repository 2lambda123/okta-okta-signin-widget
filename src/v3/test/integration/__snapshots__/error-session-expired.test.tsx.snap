// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error-session-expired should render form 1`] = `
<div>
  <span>
    <main
      class="ods-4o5zYQ ods-76eppy ods-66Tc65 ods-6lar65 ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs auth-container main-container mainViewContainer"
      id="okta-sign-in"
    >
      <div
        class="ods-4o5zYQ ods-2Se4oq ods-2VFzyK ods-1qUaxm ods-1oujX1 ods-78Q1Pv ods-3x5mmc ods-7mcS6o ods-7lpPkQ ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs siwContainer"
      >
        <div
          class="okta-sign-in-header auth-header siwHeader"
        >
          <h1
            class="MuiTypography-root MuiTypography-h1 emotion-0"
          />
        </div>
        <div
          class="MuiBox-root emotion-1"
        >
          <form
            class="o-form"
            data-se="form"
            novalidate=""
          >
            <div
              class="MuiBox-root emotion-2"
            >
              <div
                class="MuiBox-root emotion-3"
              >
                <div
                  class="MuiBox-root emotion-4"
                >
                  <div
                    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-infoboxError MuiAlert-infobox emotion-5"
                    role="alert"
                  >
                    <div
                      class="MuiAlert-icon emotion-6"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-7"
                        data-testid="ErrorOutlineIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                        />
                      </svg>
                    </div>
                    <div
                      class="MuiAlert-message emotion-8"
                    >
                      You have been logged out due to inactivity. Refresh or return to the sign in screen.
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="MuiBox-root emotion-3"
              >
                <div
                  class="MuiBox-root emotion-10"
                >
                  <a
                    class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways emotion-11"
                    href="/"
                  >
                    Back to sign in
                  </a>
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </main>
  </span>
</div>
`;
